File: js/domain/controller/Sega315_5297.js
/* 
 * Author: Enrique González Gutiérrez
 * 
 * Domain Layer: Sega 315-5297 I/O Controller Chip
 * 
 * Emulates the physical Sega 315-5297 input/output chip used in the Master System.
 * It manages the active-low voltage states of the DB-9 Controller Ports, exposing
 * them through virtual hardware registers mapped to I/O ports 0xDC and 0xDD.
 */

// Hardware active-low pin bitmasks for DB-9 Controller Port 1 (mapped to Port 0xDC)
const SEGA_IO_PIN_MASK = {
    PORT_1_UP:       0x01, // Bit 0: Controller Port 1 Up
    PORT_1_DOWN:     0x02, // Bit 1: Controller Port 1 Down
    PORT_1_LEFT:     0x04, // Bit 2: Controller Port 1 Left
    PORT_1_RIGHT:    0x08, // Bit 3: Controller Port 1 Right
    PORT_1_BUTTON_1: 0x10, // Bit 4: Controller Port 1 Button 1 / TL (Active-Low)
    PORT_1_BUTTON_2: 0x20, // Bit 5: Controller Port 1 Button 2 / TR (Active-Low)
    PORT_2_UP:       0x40, // Bit 6: Controller Port 2 Up
    PORT_2_DOWN:     0x80  // Bit 7: Controller Port 2 Down
};

class Sega315_5297 {
    constructor() {
        // Registers initialize to 0xFF (pull-up resistors mean 5V/Inactive when no buttons are pressed)
        this.portRegDC = 0xff;
        this.portRegDD = 0xff; 
    }

    /**
     * Set state of a physical pin on the controller ports.
     * Emulates active-low logic: pressing a button grounds the pin (voltage drops to 0).
     */
    writePinState(pin, isPressed) {
        if (isPressed) {
            this.portRegDC &= ~SEGA_IO_PIN_MASK[pin]; // Drop voltage to 0 (Active)
        } else {
            this.portRegDC |= SEGA_IO_PIN_MASK[pin];  // Pull-up to 1 (Inactive)
        }
    }

    /**
     * Reads register 0xDC (Port A/B inputs)
     */
    readRegisterDC() {
        return this.portRegDC;
    }

    /**
     * Reads register 0xDD (Port B / Misc inputs)
     */
    readRegisterDD() {
        return this.portRegDD;
    }

    // ------------------------------------------------------------------------
    // BACKWARD COMPATIBILITY ALIASES (For browser keyboard handlers)
    // ------------------------------------------------------------------------
    pressButton1()   { this.writePinState('PORT_1_BUTTON_1', true); }
    depressButton1() { this.writePinState('PORT_1_BUTTON_1', false); }
    pressButton2()   { this.writePinState('PORT_1_BUTTON_2', true); }
    depressButton2() { this.writePinState('PORT_1_BUTTON_2', false); }
    pressUp()        { this.writePinState('PORT_1_UP', true); }
    depressUp()      { this.writePinState('PORT_1_UP', false); }
    pressDown()      { this.writePinState('PORT_1_DOWN', true); }
    depressDown()    { this.writePinState('PORT_1_DOWN', false); }
    pressLeft()      { this.writePinState('PORT_1_LEFT', true); }
    depressLeft()    { this.writePinState('PORT_1_LEFT', false); }
    pressRight()     { this.writePinState('PORT_1_RIGHT', true); }
    depressRight()   { this.writePinState('PORT_1_RIGHT', false); }
}

File: src/domain/Sega315_5297.cpp
/**
 * Project: EGGStation - Sega & SNES Multi-System Emulator
 * Author: Enrique González Gutiérrez
 * File: src/domain/Sega315_5297.cpp
 * 
 * Domain Layer: Sega 315-5297 Input/Output Chip
 * 
 * Role:
 * Implementation of the Sega 315-5297 controller interface chip.
 * Processes the raw bitwise operations required to emulate physical active-low 
 * electrical circuits, keeping this mathematical logic entirely isolated 
 * from the JavaScript execution environment.
 */

#include "Sega315_5297.h"

Sega315_5297::Sega315_5297() {
    initialize();
}

void Sega315_5297::initialize() {
    // By default, the physical hardware utilizes VCC pull-up resistors.
    // Therefore, open circuits (unpressed buttons) are read as logical 1s (0xFF).
    portRegisterDC = 0xFF;
    portRegisterDD = 0xFF;
}

void Sega315_5297::writePinStateDC(uint8_t mask, bool isPressed) {
    if (isPressed) {
        // Active-low logic: Pressing a button pulls the pin down to Ground (0)
        portRegisterDC &= ~mask; 
    } else {
        // Releasing the button allows the pull-up resistor to pull it back high (1)
        portRegisterDC |= mask;  
    }
}

void Sega315_5297::writePinStateDD(uint8_t mask, bool isPressed) {
    if (isPressed) {
        portRegisterDD &= ~mask;
    } else {
        portRegisterDD |= mask;
    }
}

uint8_t Sega315_5297::readRegisterDC() const {
    return portRegisterDC;
}

uint8_t Sega315_5297::readRegisterDD() const {
    return portRegisterDD;
}

void Sega315_5297::restoreState(uint8_t dc, uint8_t dd) {
    portRegisterDC = dc;
    portRegisterDD = dd;
}